lib/rhc/commands/port_forward.rb

```diff
@@ -73,55 +74,77 @@ class PortForward < Base
+          list_ports_cmd = "rhc-list-ports#{options.gear ? ' --exclude-remote' : ''}"
+          ssh_cmd = "#{ssh_executable} #{ssh_uri.user}@#{ssh_uri.host} '#{list_ports_cmd} 2>&1'"
+          status, output = exec(ssh_cmd)
+          if status != 0
+            raise RHC::SSHCommandFailed.new(status, output)
+          end
+        end
 
-          if forwarding_specs.length == 0
-            # check if the gears have been stopped
-            if rest_app.gear_groups.all?{ |gg| gg.gears.all?{ |g| g["state"] == "stopped" } }
-              warn "none"
-              error "The application is stopped. Please restart the application and try again."
-              return 1
-            else
-              warn "none"
-              raise RHC::NoPortsToForwardException.new "There are no available ports to forward for this application. Your application may be stopped or idled."
-            end
+        output.each_line do |line|
+          line.chomp!
+          # FIXME: This is really brittle; there must be a better way
+          # for the server to tell us that permission (what permission?)
+          # is denied.
+          raise RHC::PermissionDeniedException.new "Permission denied." if line =~ /permission denied/i
+          # ...and also which services are available for the application
+          # for us to forward ports for.
+          if line =~ /\A\s*(\S+) -> #{HOST_AND_PORT}\z/ and (options.service.nil? or options.service.empty? or options.service.split(',').include? $1)
+            debug "Found service #{$1} with remote host #{$2} and port #{$3}"
+            fs = ForwardingSpec.new($1, $2, $3.to_i)
+            forwarding_specs << fs
+          else
+            debug line
           end
+        end
 
-          success "done"
+        if forwarding_specs.length == 0
+          # check if the gears have been stopped
+          if rest_app.gear_groups.all?{ |gg| gg.gears.all?{ |g| g["state"] == "stopped" } }
+            warn "none"
+            error "The application is stopped. Please restart the application and try again."
+            return 1
+          else
+            warn "none"
+            raise RHC::NoPortsToForwardException.new "There are no available ports to forward for this application. Your application may be stopped or idled."
+          end
+        end
 
-          begin
+        success "done"
+
+        begin
+          # if an ssh executable was specified, provide the command that be can run, assuming that the
+          # ssh executable's flags and options are the same as openssh's flags
+          if options.ssh
+            ssh_cmd_arg = forwarding_specs.map { |fs| fs.to_cmd_arg }.join(" ")
+            ssh_cmd = "#{ssh_executable} -N #{ssh_cmd_arg} #{ssh_uri.user}@#{ssh_uri.host}"
+            warn "You can try forwarding ports manually by running the command:\n  #{ssh_cmd}"
+            raise RHC::ArgumentNotValid.new("A SSH executable is specified and cannot be used with this command.")
```

Comment:
I don't think there is much we can do to improve this. I can raise something the outer rescue block will catch, but then it will not output the message in the error unless the debug flag is specified:
```
# rhc port-forward scaletest
Checking available ports ... done
Error trying to forward ports. You can try to forward manually by running:
/usr/bin/ssh -N -L 8080:127.1.244.130:8080 -L 8080:127.1.244.131:8080 -L 8080:127.1.244.129:8080 -L 20332:127.1.244.129:20332
564100f01cbe7428cf000001@scaletest-admin.dev.rhcloud.com
```

We need to be able to inform the user that the reason we were unable to forward the ports is because they specified an external ssh executable. We could change the output to always include the message from the exception, but wouldn't that be somewhat risky? The output has been this way for quite some time, we don't want to break someone's automation scripts.

Furthermore, (assuming we don't raise something the output rescue block can catch) defining `ssh_cmd` cannot be moved since it is created depending on which ports failed to forward. Some may have succeeded. The `ssh_cmd` I create includes all.